rmldatatfbbox/rmldatatfbbox/core.py

```diff
@@ -5,77 +5,158 @@
     
-#     Also suppresses warning outputs from the TF OD API.
-#     """
-#     # create a text trap and redirect stdout
-#     # to suppress printed warnings from object detection and tf
-#     text_trap = io.StringIO()
-#     sys.stdout = text_trap
-#     sys.stderr = text_trap
+    # Transformation functionality is not currently supported, thus the empty array
+    transform_metadata = []
+
+    # Variable type validation is currently not supported
+    dataset_name = config.get("dataset_name") if config.get("dataset_name") else user_input(message="What would you like to name this dataset?")
     
-#     # Calls to _dynamic_import below map to the following standard imports:
-#     #
-#     # import tensorflow as tf
-#     # from object_detection import model_hparams
-#     # from object_detection import model_lib
-#     _dynamic_import('tensorflow', 'tf')
-#     _dynamic_import('object_detection.model_hparams', 'model_hparams')
-#     _dynamic_import('object_detection.model_lib', 'model_lib')
-#     _dynamic_import('object_detection.exporter', 'exporter')
-#     _dynamic_import('object_detection.protos', 'pipeline_pb2', asfunction=True)
+    k_folds_specified = config.get("kfolds") if config.get("kfolds") else user_input(message="How many folds would you like the dataset to have?",default="5")
+
+    test_percent = config.get("test_percent") if config.get("test_percent") else .2
     
-#     # now restore stdout function
-#     sys.stdout = sys.__stdout__
-#     sys.stderr = sys.__stderr__
+    imageset_data = create.config.get("metadata").get("imageset_data")
+    labeled_images, label_to_int_dict = construct_all(imageset_data.get("image_ids"), imageset_data.get("temp_dir"))
+        
+    associated_files = {
+        "image_type_1": ".png",
+        "image_type_2": ".jpg",
+        "image_type_3": ".jpeg",
+        "metadata": ".json",
+        "labels": ".csv",
+        "PASCAL_VOC_labels": ".xml"
+    }
+
+    related_data_prefixes = {
+        'images': 'image_',
+        'labels': 'bboxLabels_'
+    }
+
+    cli_spinner("Writing out dataset locally...", write_dataset,list(labeled_images.values()),
+                                                                     custom_dataset_name=dataset_name,
+                                                                     num_folds=int(k_folds_specified),
+                                                                     test_percent=test_percent,
+                                                                     out_dir=base_dir,
+                                                                     associated_files=associated_files,
+                                                                     related_data_prefixes=related_data_prefixes,
+                                                                     label_to_int_dict=label_to_int_dict,
+                                                                     export_function=export_as_TFExample)
+
```

Comment:
The last parameter "export_function" is passing the specific function that writes the tfrecords, could be a cleaner solution (extracting the function that uses it from raven core back to plugin)

Reply:
Just an idea I had: what if instead of these single ugly functions with like 10 arguments, we went partially back to the object-oriented paradigm by creating something like a DatasetWriter class in the core. It could handle any universal stuff (e.g. dataset name, user, comments, training type, maybe even filters and transforms) so the plugin didn't have to worry about it at all. It could also have default/convenience implementations for some more common functionality, like the `label_to_int_dict`. Plus, it could take advantage of `self` to store stuff like the `label_to_int_dict` and all the metadata without having to pass them around in big dictionaries.

Then, plugins could subclass DatasetWriter and override as much functionality as they wanted. Most plugins would probably only need to override methods akin to `construct_all` and `export_as_TFExample`, and the rest would be handled for them. However, if a plugin really needed to, it could override as much functionality as it wanted, even `write_dataset` if it wanted to do something completely custom.

Basically, this is similar to what old Jigsaw did except without the weird class-level metadata and more flexible. I think this could be a bit cleaner while keeping even more unnecessary details out of the plugin. It could also be a starting point to split up big monolithic functions like `write_dataset` so that the plugin can override specific bits of functionality in an even more modular fashion.

Reply:
class based could be good, nothing inherently wrong with OO stuff. I know this PR isn't for the core but I looked at `write_dataset` and we definitely need to make some changes to make this fully extensible, that function is still calling other functions that are specific to TF. We should extract out TF related functionality to a module in the core and let plugins manually specify use of those common tools.